### PR TITLE
Stop considering `RUBY_PATCHLEVEL` for resolution

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
## What was the end-user or developer problem that led to this PR?

Bundler still has some code that considers `RUBY_PATCHLEVEL` during resolution. However, this number is no longer meaningful since Ruby 2.1 when Ruby started release patch level versions. Now the `RUBY_VERSION` properly identifies a given Ruby release so it's no longer necessary to consider the patchlevel.

This generally doesn't cause any issues, but the patchlevel does leak to some error messages, which makes them quite confusing. For example, if you set `ruby "3.1.1"` in your `Gemfile`, you may end up with an error message including the following

```
Bundler found conflicting requirements for the Ruby version:
  In Gemfile:
    Ruby (~> 3.1.1.0)

  ...
```

## What is your fix for the problem, implemented in this PR?

My fix is to stop considering patchlevels for modern rubies. With this patch, the above error now reads better

```
Bundler found conflicting requirements for the Ruby version:
  In Gemfile:
    Ruby (= 3.1.1)

  ...
``` 

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Review by indirect — APPROVED** [review]
Nice work! This seems like a good change. And cleaning up code that only matters on Ruby versions we have stopped supporting always seems worthwhile. :)

**Comment by deivid-rodriguez** [comment]
Thanks @indirect, it was a bit tricky to make this backwards compatible, but it should be ok now. I extracted some changes to a separate PR to RubyGems at #5473. I'll rebase this one after merging #5473 and CI should be fine.

**Comment by deivid-rodriguez** [comment]
Nevermind, I need to introduce changes in RubyGems and Bundler at the same time, so I closed the other PR. Will merge after CI.

**Pull request merged**
